### PR TITLE
Adding Java SDK Sample Client and Scripts installing/running Conjur and JAVA SDK Client on OpenShift

### PR DESCRIPTION
**Pull request by oburstein-hub**
The code introduced contains scripts for installing conjur on OCP and running java SDK test on it

### CODE REVIEW

**Comment by oburstein-hub** [comment]
#### What does this PR do? Is there any background context you want to provide?
This PR provides following functionality:
1. Providing sample JAVA Application which connects to Conjur and retrieves secret
2. Scripts for building, installing and running this Java Application on OpenShift
Scripts above include installation of Conjur itself on existing openshift
#### What ticket does this PR close?
Connects to [relevant GitHub issues, eg #76]
NA
#### Where should the reviewer start? How should this functionality be validated?
The reviewer should start on JavaClient class and review all the java sources and then go to scripts
The functionality can be validated by running build.sh of java and then installers
#### Screenshots (if appropriate)
#### Links to open issues for related documentation (in READMEs, docs, etc)
Connects to [relevant GitHub issues, eg cyberark/conjur-docs#76]

File: demos/open-shift-install/policy/policy1.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Filename: policy_for_human_users.yml
+# initializes users
+# bob - devops admin
+
+- !group devops
```
Comment (rafis3) [inline demos/open-shift-install/policy/policy1.yaml]: What is the purpose of this policy? It doesn't seem that you are using `bob` for anything

File: demos/open-shift-install/installer.sh [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	openssl x509 -outform der -in conjur-cert.pem -out conjur-cert.crt
+	keytool -import -alias ca -keystore trust-store.jks -storepass TrustStore -trustcacerts -file conjur-cert.crt <<< "yes
+" &> /dev/null
+	mkdir store
+	mv trust-store.jks store/.
+}
+
+usage()
+{
+    echo "usage: installer [[[--ocp-url url ] [--with-config] [--project-name project] [--account-name account] [--authenticator authenticator]] | [-h]]"
```
Comment (rafis3) [inline demos/open-shift-install/installer.sh]: Can you please explain what each flag is used for?

File: demos/java-api-client/src/main/java/myclient/JavaClient.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+public class JavaClient {
+
+    public static void main(String args[]) {
+        System.out.println("Running Conjur Java SDK Example");
+        String truststoreFileName = "/run/conjur/store/trust-store.jks";
+        try {
+            while (true)
+            {
+                System.out.println("Check if " + truststoreFileName + "file appeared");
```
Comment (rafis3) [inline demos/java-api-client/src/main/java/myclient/JavaClient.java]: You are missing a space after the file name. Also I think it would be helpful to explain what this file is. Why is the code waiting for it?

File: demos/open-shift-install/java-client-installer.sh [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        env:
+          - name: CONJUR_AUTHN_URL
+            value: https://conjur-oss.$PROJECT_NAME.svc.cluster.local/authn-k8s/$AUTHENTICATOR
+          - name: CONJUR_APPLIANCE_URL
+            value: https://conjur-oss.$PROJECT_NAME.svc.cluster.local
+          - name: CONJUR_ACCOUNT
+            value: $ACCOUNT_NAME
+          - name: CONJUR_AUTHN_LOGIN
+            value: admin
+          - name: PATH_TO_AUTH_TOKEN_FILE
```
Comment (rafis3) [inline demos/open-shift-install/java-client-installer.sh]: Please align to our existing standard and change the environment variable name to `CONJUR_AUTHN_TOKEN_FILE`

File: demos/java-api-client/src/main/java/myclient/JavaClient.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            System.setProperty("CONJUR_AUTHN_LOGIN", System.getenv("CONJUR_AUTHN_LOGIN"));
+            System.setProperty("CONJUR_APPLIANCE_URL", System.getenv("CONJUR_APPLIANCE_URL"));
+            //For OpenShift
+
+            System.out.println("PATH_TO_AUTH_TOKEN_FILE = " + System.getenv("PATH_TO_AUTH_TOKEN_FILE"));
+            Token token = Token.fromFile(Paths.get(System.getenv("PATH_TO_AUTH_TOKEN_FILE")));
+            System.out.println("Create Conjur API Instance");
+            Conjur conjur = new Conjur(token);
+            String secret = conjur.variables().retrieveSecret("variables/mypassword");
+            System.out.println("retrieved secret = " + secret);
```
Comment (rafis3) [inline demos/java-api-client/src/main/java/myclient/JavaClient.java]:
```suggestion
            System.out.println("Retrieved secret = " + secret);
```

File: demos/java-api-client/src/main/java/myclient/JavaClient.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        catch (Exception e)
+        {
+            System.out.println("Exception:" + e);
```
Comment (rafis3) [inline demos/java-api-client/src/main/java/myclient/JavaClient.java]: Do you feel that printing the `Exception` variable is enough? Maybe we should print the message? What information do I currently get from this at the moment?

File: demos/open-shift-install/installer.sh [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+	oc delete ClusterRoleBinding conjur-oss-conjur-authenticator
+
+	echo "authenticators: \"authn-k8s/$AUTHENTICATOR,authn\"
+dataKey: \"$DATA_KEY\"
+account: \"$ACCOUNT_NAME\"
+
+image:
+  repository: \"cyberark/conjur\"
+  tag: \"latest\"
+  pullPolicy: Always
```
Comment (rafis3) [inline demos/open-shift-install/installer.sh]: Usually it's safer to use `IfNotPresent` which will tell OpenShift to pull the image only if it doesn't exist locally

File: demos/open-shift-install/installer.sh [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+	oc exec -it "$CONJUR_CLI_POD" conjur init <<< "https://conjur-oss
+yes
+$ACCOUNT_NAME
+y
+"
+
+	oc exec -it "$CONJUR_CLI_POD" conjur authn login < policy/authnInput
+	
+	oc exec -it "$CONJUR_CLI_POD" conjur policy load root policy/policy1.yaml
```
Comment (rafis3) [inline demos/open-shift-install/installer.sh]: Can you please give useful names for the policy? Why do you have 4? What is the purpose of each of them? This purpose can be reflected in the policy name

File: demos/java-api-client/src/main/java/myclient/JavaClient.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import net.conjur.api.Token;
+
+import java.io.File;
+import java.nio.file.Paths;
+
+public class JavaClient {
+
+    public static void main(String args[]) {
+        System.out.println("Running Conjur Java SDK Example");
+        String truststoreFileName = "/run/conjur/store/trust-store.jks";
```
Comment (rafis3) [inline demos/java-api-client/src/main/java/myclient/JavaClient.java]: Are you using a hard coded KeyStore that is saved in this repository? If so, I would suggest generating a new one and use it instead.

File: demos/open-shift-install/installer.sh [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+dataKey: \"$DATA_KEY\"
+account: \"$ACCOUNT_NAME\"
+
+image:
+  repository: \"cyberark/conjur\"
+  tag: \"latest\"
+  pullPolicy: Always
+
+ssl:
+  hositname: custom.domainname.com
```
Comment (rafis3) [inline demos/open-shift-install/installer.sh]: It should be `hostname`

File: demos/open-shift-install/java-client-installer.sh [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+DEPLOYMENT_NAME=conjur-java-api-example
+
+function install {
+
+	
+##	oc login $OPENSHIFT_URL
+	echo "entering project $PROJECT_NAME"
+	oc project $PROJECT_NAME
+
+        CONJUR_OSS_POD=$( echo "$CONJUR_OSS_POD_LINE" | awk '{print $1}' )
```
Comment (rafis3) [inline demos/open-shift-install/java-client-installer.sh]: Please fix the indention

File: demos/open-shift-install/java-client-installer.sh [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+        oc get pods
```
Comment (rafis3) [inline demos/open-shift-install/java-client-installer.sh]: Why is this needed? Maybe if this output has a purpose, you could add a matching message with it?

File: demos/open-shift-install/java-client-installer.sh [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+        TOKEN=$( oc whoami -t )
+
+        docker tag conjur-java-client:latest $DOCKER_URL/$PROJECT_NAME/conjur-java-client:latest
```
Comment (rafis3) [inline demos/open-shift-install/java-client-installer.sh]: Can you please add some messages for each section that explains the flow to the user? So that before each block of actions, like uploading the image, deploying Conjur, loading policies, will have matching messages during the script run

File: demos/java-api-client/ConjurJavaAPI.iml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="JAVA_MODULE" version="4" />
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/ConjurJavaAPI.iml]: done

File: demos/java-api-client/pom.xml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                        <configuration>
+                            <shadedClassifierName>with-dependencies</shadedClassifierName>
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/pom.xml]: in order to distinguish between fat and regular jars. "with-dependencies" is a fat jar

File: demos/java-api-client/pom.xml [hunk 16]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/pom.xml]: done

File: demos/java-api-client/src/main/java/myclient/JavaClient.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        try {
+            while (true)
+            {
+                System.out.println("Check if " + truststoreFileName + "file appeared");
+                File tempFile = new File(truststoreFileName);
+
+                if (tempFile.exists())
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/src/main/java/myclient/JavaClient.java]: done

File: demos/java-api-client/build.sh [hunk 18]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+rm -rf target
+
+git clone https://github.com/cyberark/conjur-api-java.git
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/build.sh]: done

File: demos/java-api-client/src/main/java/myclient/JavaClient.java (hunk 9 above)
Comment (oburstein-hub) [reply demos/java-api-client/src/main/java/myclient/JavaClient.java]: it is not hard-coded. It is generated via installation.sh script

File: demos/open-shift-install/installer.sh (hunk 10 above)
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: done

File: demos/open-shift-install/java-client-installer.sh (hunk 11 above)
Comment (oburstein-hub) [reply demos/open-shift-install/java-client-installer.sh]: was OK in vi. Tried to make it look OK in the above viewer either. Hopefully it helps

File: demos/open-shift-install/java-client-installer.sh (hunk 12 above)
Comment (oburstein-hub) [reply demos/open-shift-install/java-client-installer.sh]: added message

File: demos/open-shift-install/java-client-installer.sh (hunk 13 above)
Comment (oburstein-hub) [reply demos/open-shift-install/java-client-installer.sh]: done

File: demos/java-api-client/build.sh [hunk 19]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+rm -rf target
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/java-api-client/build.sh]: done

File: demos/open-shift-install/java-client-installer.sh [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+OPENSHIFT_URL=
+PROJECT_NAME=
+ACCOUNT_NAME=myaccount
+AUTHENTICATOR=myauthenticator
+DEPLOYMENT_NAME=conjur-java-api-example
+
+function install {
+
+	
+##	oc login $OPENSHIFT_URL
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/open-shift-install/java-client-installer.sh]: done

File: demos/open-shift-install/installer.sh [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+fi
+
+rm -rf conjur-cert.crt
+rm -rf conjur-cert.pem
+rm -rf custom-values.yaml
+rm -rf policy/authnInput
+rm -rf policy/policy2.yaml
+rm -rf policy/policy3.yaml
+rm -rf policy/policy4.yaml
+##echo "Now run the following commands on MAC:"
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: removed

File: demos/open-shift-install/installer.sh [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+
+#set -x
+
+OPENSHIFT_URL=
+PROJECT_NAME=
+ACCOUNT_NAME=myaccount
+AUTHENTICATOR=myauthenticator
+
+function install {
+
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: done

File: demos/open-shift-install/installer.sh [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+		echo "Waiting for conjur pod to be up..."
+		sleep 2
+	done
+
+	oc get pods
+
+	if [ "$CONTAINERS_STATUS" != "2/2" ]; then
+                echo "Conjur POD did not come up properly - exiting"
+		exit 1
+        fi
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: it was matched for vi - but not for git viewer. Now after substitution of tab by 2 spaces hopefully it will be OK

File: demos/open-shift-install/installer.sh [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+	echo "Installation done"
+	oc adm policy add-scc-to-user anyuid "system:serviceaccount:$PROJECT_NAME:default"
+	} &> /dev/null
+
+	CONJUR_OSS_POD_LINE=$( oc get pods | grep conjur-oss | (head -n1 && tail -n1) )
+
+	CONJUR_OSS_POD=$( echo "$CONJUR_OSS_POD_LINE" | awk '{print $1}' )
+
+	for i in {1..50}
+	do
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: done

File: demos/open-shift-install/policy/policy1.yaml (hunk 1 above)
Comment (oburstein-hub) [reply demos/open-shift-install/policy/policy1.yaml]: removed bob - it was just a sample user

File: demos/open-shift-install/installer.sh (hunk 2 above)
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: do you want me to add explanation in README or here ?

File: demos/java-api-client/src/main/java/myclient/JavaClient.java (hunk 3 above)
Comment (oburstein-hub) [reply demos/java-api-client/src/main/java/myclient/JavaClient.java]: explanation added as comment in java client code. It it OK ?

File: demos/open-shift-install/java-client-installer.sh (hunk 4 above)
Comment (oburstein-hub) [reply demos/open-shift-install/java-client-installer.sh]: done

File: demos/java-api-client/src/main/java/myclient/JavaClient.java (hunk 5 above)
Comment (oburstein-hub) [reply demos/java-api-client/src/main/java/myclient/JavaClient.java]: done

File: demos/java-api-client/src/main/java/myclient/JavaClient.java (hunk 6 above)
Comment (oburstein-hub) [reply demos/java-api-client/src/main/java/myclient/JavaClient.java]: We will get stuck trace and message - it is enough ?

File: demos/open-shift-install/installer.sh (hunk 7 above)
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: I used to work this way but it might be a problem if the existing image is damaged
I would prefer to leave it like this for the example -  just to increase chances that it will run OK

File: demos/open-shift-install/installer.sh (hunk 8 above)
Comment (oburstein-hub) [reply demos/open-shift-install/installer.sh]: done

**Comment by oburstein-hub** [comment]
closed because fixes are done I I will open a new one
